### PR TITLE
Add handler function for RADIUS attributes in CoA messages

### PR DESCRIPTION
**Pull request by rr-santos**
This commit adds the handle_coa_attributes function in the dm_coa.c
file. This function is resposible for iterating through the radius
attributes from a CoA request message and update the corresponding
session data structure values.

This allows the internal session data structure to be updated whenever a
CoA request contains new values to the existing attributes. In addition,
by calling the handle_coa_attributes function before the ev_radius_coa
event is fired, the redis module can publish this event with the new
updated session attributes.

Signed-off-by: Ricardo Santos <ricardo.santos@bisdn.de>

### CODE REVIEW

**Comment by butjar** [comment]
For reference, to enable the publishing of CoA messages you need to enable `ev_radius_coa` in the config file:

```
[redis]
...
ev_radius_coa=yes
```

**Comment by butjar** [comment]
# Testing instructions
## Prerequisites
 
1. This guide assumes a RADIUS server (we use [freeradius](https://freeradius.org/)) is up and running. We use [FreeRADIUS SQL Module (`rlm_sql`)](https://wiki.freeradius.org/modules/rlm_sql#sql-schema-and-usage_radgroupreply-table). Furthermore, we have configured `freeradius` to return `Filter-Id`s for different user groups
2. `redis` installed (e.g. on Ubuntu `sudo apt install redis-server`) 
3. `pppd` Point-to-Point Protocol Daemon installed (e.g. on Ubuntu `sudo apt install ppp`)
4. `radclient` RADIUS client CLI installed (e.g. on Ubuntu `sudo apt install freeradius`)

## Setup

1. Create veth pair to emulate the link between `accel-ppp` and the client isolated in a `netns`: 
    ```
    sudo ip link add veth_accel type veth peer name veth_client
    sudo ip link set veth_accel up
    sudo ip netns add client
    sudo ip link set veth_client netns client
    sudo ip netns exec client ip link set veth_client up
    ```
2. Checkout this feature branch:
    ```
    git checkout handle_coa_attributes
    ```
3. Build and install `accel-ppp`:
    ```
    mkdir build
    cd build
    cmake -DBUILD_DRIVER=FALSE -DRADIUS=TRUE -DREDIS=TRUE ..
    sudo make install
    ```
4. Create `accel-ppp.conf` with the following content (replace `<RADIUS_IP>` and `<ACCEL_PPP_ADDRESS>` in the `[radius]` section):
    ```
    [modules]
    log_file
    redis
    pptp
    l2tp
    pppoe

    auth_mschap_v2
    auth_mschap_v1
    auth_chap_md5
    auth_pap

    radius

    [core]
    log-error=/dev/stderr
    thread-count=4

    [ppp]
    verbose=1
    min-mtu=1280
    mtu=1400
    mru=1400
    ipv4=require
    ipv6=deny
    lcp-echo-interval=20
    lcp-echo-timeout=120
    unit-cache=1
    unit-preallocate=1

    [auth]
    noauth=0

    [pptp]
    verbose=1

    [pppoe]
    verbose=1
    called-sid=mac
    tr101=1
    interface=veth_accel

    [radius]
    gw-ip-address=<RADIUS_IP>
    nas-ip-address=<ACCEL_PPP_ADDRESS>
    nas-port=ifname
    nas-identifier=nas1
    server=<RADIUS_IP>,SECRET,auth-port=1812,acct-port=1813,req-limit=50,fail-timeout=0,max-fail=10,weight=1
    dae-server=127.0.0.1:1700,SECRET
    verbose=1

    [log]
    log-file=/dev/stdout
    log-emerg=/dev/stderr
    log-fail-file=/dev/stdout
    log-debug=/dev/stdout
    copy=1
    color=1
    level=3

    [cli]
    verbose=1
    telnet=127.0.0.1:2000
    tcp=127.0.0.1:2001

    [redis]
    host=localhost
    port=6379
    pubchan=accel-ppp

    # select the event types to emit a message via redis
    ev_ses_acct_start=yes
    ev_ses_pre_finished=yes
    ev_radius_coa=yes
    ```
5. Start `redis-server` in a different terminal window:
    ```
    redis-server
    ```
6. Monitor `redis` in another terminal window:
    ```
    redis-cli monitor
    ```
7. Start `accel-ppp` (still from the `build` folder):
    ```
    sudo ./accel-pppd/accel-pppd -c accel-ppp.conf
    ```
8. Enter the `client` netns and create a client session in another window (replace the exported variables):
    ```
    sudo ip netns exec client /bin/bash
    ```
    ```
    export IFACE="veth_client" MTU="1350" MRU="1350" PPP_USER="sp-bronze" PPP_PASSWORD="secret"
    sudo pppd pty "pppoe -I $IFACE -T 80 -U -m $MTU" \
    noccp \
    ipparam $IFACE \
    linkname $IFACE \
    noipdefault \
    noauth \
    default-asyncmap \
    defaultroute \
    hide-password \
    updetach \
    mtu $MTU \
    mru $MRU \
    noaccomp \
    nodeflate \
    nopcomp \
    novj \
    novjccomp \
    lcp-echo-interval 40 \
    lcp-echo-failure 3 \
    user $PPP_USER \
    password $PPP_PASSWORD
    ```
    Remember the clients `Acct-Session-Id` from the `accel-ppp` logs on stdout (you can also see it in the `redis-cli monitor`). We will use it, to identify our "user" in the CoA request. E.g.:
    ```
    [2021-06-08 10:30:16]:  info: ppp0: sp-bronze: authentication succeeded
    [2021-06-08 10:30:16.163] ppp0: d89da8604fc1b693: send [RADIUS(1) Accounting-Request id=1 <User-Name "sp-bronze"> 
    <NAS-Identifier "nas1"> <NAS-IP-Address 172.20.0.18> <NAS-Port 0> <NAS-Port-Id "ppp0"> <NAS-Port-Type Virtual> 
    <Service-Type Framed-User> <Framed-Protocol PPP> <Calling-Station-Id "92:75:74:70:aa:8c"> <Called-Station-Id 
    "52:61:c1:26:73:cc"> <Acct-Status-Type Start> <Acct-Authentic RADIUS> <Acct-Session-Id "d89da8604fc1b693"> <Acct- 
   Session-Time 0> <Acct-Input-Octets 0> <Acct-Output-Octets 0> <Acct-Input-Packets 0> <Acct-Output-Packets 0> <Acct- 
   Input-Gigawords 0> <Acct-Output-Gigawords 0> <Framed-IP-Address 192.51.0.2>]
    ```

## Test cases
### Update the `Filter-id` of the created user
#### Test steps
0. Setup the session as described in the *Setup* section
1. Send a CoA message to update the `Filter-Id` (the `Acct-Session-Id` must match the one of the created session):
    ```
    echo "Acct-Session-Id=d89da8604fc1b693,Filter-Id=2/100000/50000" | radclient -x localhost:1700 coa 'SECRET'
    ```

#### Expected Behavior

1. Check the response. A `CoA-ACK` was returned:
    ```
    Sent CoA-Request Id 120 from 0.0.0.0:40251 to 127.0.0.1:1700 length 54
            Acct-Session-Id = "d89da8604fc1b693"
            Filter-Id = "2/100000/50000"
    Received CoA-ACK Id 120 from 127.0.0.1:1700 to 0.0.0.0:0 length 20
    ```
2. Look at the output of the `redis-cli monitor` terminal. A CoA event was published with the updated `Filter-Id` value as `qos_val`:
    ```
    1623141295.270437 [0 127.0.0.1:38296] "PUBLISH" "accel-ppp" "{ \"event\": \"coa\", \"ctrl_type\": \"pppoe\", 
    \"channel_name\": \"92:75:74:70:aa:8c\", \"session_id\": \"d89da8604fc1b693\", \"called_station_id\": \"52:61:c1:26:73:cc\", 
    \"calling_station_id\": \"92:75:74:70:aa:8c\", \"name\": \"pppoe\", \"username\": \"sp-bronze\", \"ip_addr\": \"192.51.0.2\", 
    \"pppoe_sessionid\": 1, \"ctrl_ifname\": \"veth_accel\", \"nas_identifier\": \"nas1\", \"qos_val\": \"2\\/100000\\/50000\" }"
    ```

### Send Disconnect Message (DM)
#### Test steps

0. Setup the session as described in the *Setup* section (or use the setup from the former test case)
1. Send a DM to wipe the user session (`Acct-Session-Id` and `User-Name` must match the ones of the created session):
    ```
   echo "Acct-Session-Id=d89da8604fc1b693,User-Name=sp-bronze" | radclient -x localhost:1700 disconnect 'SECRET'
   ```

#### Expected Behavior

1. Check the response. A `Disconnect-ACK` was returned:
    ```
    Sent Disconnect-Request Id 249 from 0.0.0.0:38597 to 127.0.0.1:1700 length 49
            Acct-Session-Id = "d89da8604fc1b693"
            User-Name = "sp-bronze"
    Received Disconnect-ACK Id 249 from 127.0.0.1:1700 to 0.0.0.0:0 length 20
    ```
2. The shell in the client window shows that a `PADT` was received:
    ```
    # PADT: Generic-Error: accel-ppp
    PADT: Generic-Error:
    ```
3. Check with `accel-cmd` that the session was wiped:
    ```
    ./build/accel-cmd/accel-cmd show sessions
    ifname | username | calling-sid | ip | type | comp | state | uptime
    --------+----------+-------------+----+------+------+-------+--------
    ```

## Cleanup

1. Delete veth pair:
    ```
    sudo ip link delete veth_accel
    ```
2. Delete the `client` netns:
    ```
    sudo ip netns delete client
    ```

File: accel-pppd/radius/dm_coa.c [hunk 1]
```diff
@@ -158,6 +179,12 @@ static void coa_request(struct radius_pd_t *rpd)
 		rad_packet_print(rpd->dm_coa_req, NULL, log_ppp_info2);
 	}
 
+	/* Process RADIUS attributes in CoA request message and update the
+	corresponding session with the new values. This is done before the triton
+	event is triggered so the redis module can read and publish the new values
+	from the session data structure */
+	handle_coa_attributes(rpd);
+
```
Comment (butjar) [inline accel-pppd/radius/dm_coa.c:187]: For reference: For [`ipoe`](https://github.com/xebd/accel-ppp/blob/a77f9e071f084918bbb805c2509aec0f47fcf95b/accel-pppd/ctrl/ipoe/ipoe.c#L2417-L2500) the radius attributes in the CoA messages are handled within the module

Comment (butjar) [reply accel-pppd/radius/dm_coa.c:187]: Probably we should at least check the `ev->ses->ctrl->type` ([similar as in ipoe.c](https://github.com/xebd/accel-ppp/blob/a77f9e071f084918bbb805c2509aec0f47fcf95b/accel-pppd/ctrl/ipoe/ipoe.c#L2425))

**Comment by butjar** [comment]
:+1: Tested with the [Test Instructions](https://github.com/dpdk-vbng-cp/accel-ppp/pull/14#issuecomment-856097104) :heavy_check_mark:

**Comment by butjar** [comment]
@rr-santos Could you please create a new image and update `latest`?

**Comment by rr-santos** [comment]
> @rr-santos Could you please create a new image and update `latest`?

On it!

**Comment by rr-santos** [comment]
https://hub.docker.com/layers/bisdn/accel-pppd/latest/images/sha256-c9fa7a2ecb4a037867160a0f083ec4b78e196174f028a8250b9a80c46e4c89b6?context=repo

**Pull request merged**